server/src/main/java/org/candlepin/async/JobMessageDispatcher.java

```diff
@@ -44,12 +49,158 @@
+
+                log.debug("Created new CPM session: {}", this.session);
+            }
+
+            return this.session;
+        }
+
+        /**
+         * Fetches the current CPM producer, creating a new one if necessary.
+         *
+         * @return
+         *  a CPMProducer instance
+         */
+        public CPMProducer getProducer() throws CPMException {
+            if (this.producer == null || this.producer.isClosed()) {
+                log.debug("Creating new CPM producer for job message dispatch for thread {}",
+                    Thread.currentThread());
+
+                CPMSession session = this.getSession();
+                CPMProducerConfig config = session.createProducerConfig();
+
+                // Add any other job-system-specific producer configuration here
+
+                this.producer = session.createProducer(config);
+
+                log.debug("Created new CPM producer: {}", this.producer);
+            }
+
+            return this.producer;
+        }
+
+        /**
+         * Closes any session resources this session store may have open
+         */
+        private void close() throws CPMException {
+            if (this.producer != null) {
+                this.producer.close();
+                this.producer = null;
+            }
+
+            if (this.session != null) {
+                this.session.close();
+                this.session = null;
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public String toString() {
+            return String.format("ThreadSessionStore [session: %s, producer: %s]",
+                this.session, this.producer);
+        }
+    }
+
+    /**
+     * The ThreadReference class is a WeakReference implementation with stable equals and hashCode
+     * methods, allowing it to be used as a key in maps and in sets.
+     */
+    private class ThreadReference extends WeakReference<Thread> {
+        private final int hashCode;
+
+        public ThreadReference(Thread thread, ReferenceQueue<? super Thread> queue) {
+            super(thread, queue);
+
+            this.hashCode = thread != null ? thread.hashCode() : 0;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean equals(Object cmp) {
+            if (cmp == this) {
+                return true;
+            }
+
+            if (cmp instanceof ThreadReference) {
+                Thread lhs = this.get();
+                Thread rhs = ((ThreadReference) cmp).get();
+
+                return lhs != null ?
+                    lhs.equals(rhs) :
+                    rhs == null && this.hashCode() == ((ThreadReference) cmp).hashCode();
+            }
+
+            return false;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public int hashCode() {
+            return this.hashCode;
+        }
+    }
+
+
+
     private final Configuration config;
```

Comment:
Not a part of your changes but there is an unused dependency.

Reply:
Fixed